libraries/AP_Math/SCurve.cpp

```diff
@@ -0,0 +1,963 @@
+        move_from_time_pos_vel_accel(get_time_elapsed() + time_to_destination / 2.0f, turn_pos, turn_vel, turn_accel);
+        next_leg.move_from_time_pos_vel_accel(time_to_destination / 2.0f, turn_pos, turn_vel, turn_accel);
+        const float speed_min = MIN(get_speed_along_track(), next_leg.get_speed_along_track());
+        const float accel_min = MIN(get_accel_along_track(), next_leg.get_accel_along_track());
+        if ((get_time_remaining() < next_leg.time_end() / 2.0f) && (turn_pos.length() < wp_radius) &&
+             (Vector2f(turn_vel.x, turn_vel.y).length() < speed_min) &&
+             (Vector2f(turn_accel.x, turn_accel.y).length() < 2.0f*accel_min)) {
+            next_leg.move_from_pos_vel_accel(dt, target_pos, target_vel, target_accel);
+        }
+    } else if (!is_zero(next_leg.get_time_elapsed())) {
+        next_leg.move_from_pos_vel_accel(dt, target_pos, target_vel, target_accel);
+        if (next_leg.get_time_elapsed() >= get_time_remaining()) {
+            s_finished = true;
+        }
+    }
+
+    return s_finished;
+}
+
+// increment time pointer and return the position, velocity and acceleration vectors relative to the origin
+void SCurve::move_from_pos_vel_accel(float dt, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    advance_time(dt);
+    float scurve_P1 = 0.0f;
+    float scurve_V1, scurve_A1, scurve_J1;
+    get_jerk_accel_vel_pos_at_time(time, scurve_J1, scurve_A1, scurve_V1, scurve_P1);
+    pos += delta_unit * scurve_P1;
+    vel += delta_unit * scurve_V1;
+    accel += delta_unit * scurve_A1;
+    position_sq = sq(scurve_P1);
+}
+
+// increment time pointer and return the position, velocity and acceleration vectors relative to the destination
+void SCurve::move_to_pos_vel_accel(float dt, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    advance_time(dt);
+    float scurve_P1 = 0.0f;
+    float scurve_V1, scurve_A1, scurve_J1;
+    get_jerk_accel_vel_pos_at_time(time, scurve_J1, scurve_A1, scurve_V1, scurve_P1);
+    pos += delta_unit * scurve_P1;
+    vel += delta_unit * scurve_V1;
+    accel += delta_unit * scurve_A1;
+    position_sq = sq(scurve_P1);
+    pos -= track;
+}
+
+// return the position, velocity and acceleration vectors relative to the origin at a specified time along the path
+void SCurve::move_from_time_pos_vel_accel(float time_now, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    float scurve_P1 = 0.0f;
+    float scurve_V1, scurve_A1, scurve_J1;
+    get_jerk_accel_vel_pos_at_time(time_now, scurve_J1, scurve_A1, scurve_V1, scurve_P1);
+    pos += delta_unit * scurve_P1;
+    vel += delta_unit * scurve_V1;
+    accel += delta_unit * scurve_A1;
+}
+
+// time has reached the end of the sequence
+bool SCurve::finished() const
+{
+    return ((time >= time_end()) || (position_sq >= track.length_squared()));
+}
+
+// magnitude of the position vector at the end of the sequence
+float SCurve::pos_end() const
+{
+    if (num_segs != segments_max) {
+        return 0.0f;
```

Comment:
Is this a "shouldn't happen"?  flow_of_control error perhaps?

Similary for the other accessors here.

Reply:
this is a bit of defensive programming but we shouldn't need yet another internal-error raised.  It will already have been raised somewhere else.